core/src/VolatilityOracle.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.17;
+
+import {IUniswapV3Pool} from "v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+
+import {MAX_SIGMA, IV_SCALE, IV_CHANGE_PER_SECOND, ORACLE_LOOKBACK, FEE_GROWTH_GLOBALS_SAMPLE_PERIOD} from "./libraries/constants/Constants.sol";
+import {Oracle} from "./libraries/Oracle.sol";
+import {Volatility} from "./libraries/Volatility.sol";
+
+/// @title VolatilityOracle
+/// @author Aloe Labs, Inc.
+/// @dev "Test everything; hold fast what is good." - 1 Thessalonians 5:21
+contract VolatilityOracle {
+    struct LastWrite {
+        uint8 index;
+        uint32 time;
+        uint216 iv;
+    }
+
+    mapping(IUniswapV3Pool => Volatility.PoolMetadata) public cachedMetadata;
+
+    mapping(IUniswapV3Pool => Volatility.FeeGrowthGlobals[60]) public feeGrowthGlobals;
+
+    mapping(IUniswapV3Pool => LastWrite) public lastWrites;
+
+    function prepare(IUniswapV3Pool pool) external {
+        cachedMetadata[pool] = _getPoolMetadata(pool);
+
+        if (lastWrites[pool].time == 0) {
+            feeGrowthGlobals[pool][0] = _getFeeGrowthGlobalsNow(pool);
+            lastWrites[pool] = LastWrite({index: 0, time: uint32(block.timestamp), iv: uint216(MAX_SIGMA)});
+        }
+    }
+
+    function update(IUniswapV3Pool pool) external returns (uint160, uint256) {
+        unchecked {
+            // Read `lastWrite` info from storage
+            LastWrite memory lastWrite = lastWrites[pool];
+
+            // If fewer than `VOLATILITY_SAMPLE_PERIOD` seconds have elapsed, return early. We
+            // still fetch the latest TWAP, but we do not sample feeGrowthGlobals or update IV
+            uint256 timeSinceLastWrite = block.timestamp - lastWrite.time;
+            if (timeSinceLastWrite < FEE_GROWTH_GLOBALS_SAMPLE_PERIOD) {
+                (uint160 sqrtMeanPriceX96, ) = Oracle.consult(pool, ORACLE_LOOKBACK);
+                return (sqrtMeanPriceX96, lastWrite.iv);
+            }
+
+            // Prepare to call all getters
+            Volatility.FeeGrowthGlobals[60] storage arr = feeGrowthGlobals[pool];
+
+            // Call all getters
+            Volatility.PoolData memory data = _getPoolData(pool);
+            Volatility.FeeGrowthGlobals memory a = _getFeeGrowthGlobalsOld(arr, lastWrite.index);
+            Volatility.FeeGrowthGlobals memory b = _getFeeGrowthGlobalsNow(pool);
+
+            // Default to using the existing IV
+            uint256 iv = lastWrite.iv;
+            // Only update IV if the feeGrowthGlobals samples are approximately 1 hour apart
+            if (
+                isInInterval({
+                    min: 1 hours - 5 minutes, // NOTE: Keeping these constants here because they're related to `arr.length`
+                    x: b.timestamp - a.timestamp,
+                    max: 1 hours + 5 minutes
+                })
+            ) {
```

Comment:
So we're updating IV if the `feeGrowthGlobals` samples are 1 hour apart (technically, we'll do an update at some time between 55 mins to 1 hour 5 mins). I also see that the length of our array is 60 (60 mins in 1 hour) which is why we're keeping this number. Why 1 hour? Even bigger question - are we guaranteed to call update periodically so that we don't miss this window? What happens if we miss this window?

Reply:
I plan to write a bot similar to the liquidation bot that will call this every hour, on the hour for popular pools. If we miss a window, it's not the end of the world though -- the IVs (and "equivalent collateral factors") would just freeze in place until we start calling it again.

Also, there's an incentive for liquidators to call it when IV goes up (since that brings people closer to liquidation) and there's an incentive for borrowers to call it when IV goes down (since that makes them healthier and/or lets them borrow more). So long-term we shouldn't have to be the only ones calling `update`

Reply:
Creating a bot for popular pools sounds good. Also, good to know that we have safety measures in case we miss the window and that calling this is aligned with the LP's incentives.